examples/step-50/step-50.cc

```diff
@@ -14,912 +14,1312 @@
+    const MatrixFree<dim, number, VectorizedArray<number>> &mf_storage) const;
+};
 
-    VectorType solution;
-    VectorType system_rhs;
 
-    const unsigned int degree;
+template <int dim>
+double Coefficient<dim>::value(const Point<dim> &p, const unsigned int) const
+{
+  for (int d = 0; d < dim; ++d)
+    {
+      if (p[d] < -0.5)
+        return 100.0;
+    }
+  return 1.0;
+}
 
-    // Finally we are storing the various parallel multigrid matrices. Our
-    // problem is self-adjoint, so the interface matrices are the transpose
-    // of each other, so we only need to compute/store them once.
-    MGLevelObject<MatrixType> mg_matrices;
-    MGLevelObject<MatrixType> mg_interface_matrices;
-    //
-    MGConstrainedDoFs mg_constrained_dofs;
-  };
 
+template <int dim>
+template <typename number>
+VectorizedArray<number>
+Coefficient<dim>::value(const Point<dim, VectorizedArray<number>> &p,
+                        const unsigned int) const
+{
+  VectorizedArray<number> return_value = VectorizedArray<number>();
+  for (unsigned int i = 0; i < VectorizedArray<number>::size(); ++i)
+    {
+      bool found = false;
+      for (int d = 0; d < dim; ++d)
+        if (p[d][i] < -0.5)
+          {
+            return_value[i] = 100.0;
+            found           = true;
+            break;
+          }
 
+      if (!found)
+        return_value[i] = 1.0;
+    }
 
-  // @sect3{Nonconstant coefficients}
+  return return_value;
+}
 
-  // The implementation of nonconstant
-  // coefficients is copied verbatim
-  // from step-5 and step-6:
 
-  template <int dim>
-  class Coefficient : public Function<dim>
-  {
-  public:
-    virtual double value(const Point<dim> & p,
-                         const unsigned int component = 0) const override;
+template <int dim>
+template <typename number>
+number Coefficient<dim>::average_value(
+  const std::vector<Point<dim, number>> &points) const
+{
+  number average(0);
+  for (unsigned int i = 0; i < points.size(); ++i)
+    average += value(points[i]);
+  average /= points.size();
 
-    virtual void value_list(const std::vector<Point<dim>> &points,
-                            std::vector<double> &          values,
-                            const unsigned int component = 0) const override;
-  };
+  return average;
+}
 
 
 
-  template <int dim>
-  double Coefficient<dim>::value(const Point<dim> &p, const unsigned int) const
-  {
-    if (p.square() < 0.5 * 0.5)
-      return 5;
-    else
-      return 1;
-  }
+template <int dim>
+template <typename number, int degree>
+std::shared_ptr<Table<2, VectorizedArray<number>>>
+Coefficient<dim>::create_coefficient_table(
+  const MatrixFree<dim, number, VectorizedArray<number>> &mf_storage) const
+{
+  std::shared_ptr<Table<2, VectorizedArray<number>>> coefficient_table;
+  coefficient_table = std::make_shared<Table<2, VectorizedArray<number>>>();
 
+  FEEvaluation<dim, degree, degree + 1, 1, number> fe_eval(mf_storage);
```

Comment:
You could avoid the template argument on this function if you set
```c++
FEEvaluation<dim, -1, 0, 1, number> fe_eval(mf_storage);
```
This would result in slower `evaluate` and `integrate` calls because everything must be done by run-time parameters, but since you only use the quadrature points it does not matter. This makes the call slightly simpler.

Reply:
I was unaware of the `fe_degree=-1` before. Thanks! This also makes the call much nicer (I can remove the `.template`)